nalgebra-macros/tests/tests.rs

```diff
@@ -305,3 +305,125 @@ fn dvector_arbitrary_expressions() {
     let a_expected = DVector::from_column_slice(&[1 + 2, 2 * 3, 4 * f(5 + 6), 7 - 8 * 9]);
     assert_eq_and_type!(a, a_expected);
 }
+
+#[test]
+fn stack_simple() {
+    let m = stack![
+        Matrix2::<usize>::identity(), 0;
+        0, &Matrix2::identity();
+    ];
+
+    assert_eq_and_type!(m, Matrix4::identity());
+}
+
+#[test]
+fn stack_diag() {
+    let m = stack![
+        0, matrix![1, 2; 3, 4;];
+        matrix![5, 6; 7, 8;], 0;
+    ];
+
+    let res = matrix![
+        0, 0, 1, 2;
+        0, 0, 3, 4;
+        5, 6, 0, 0;
+        7, 8, 0, 0;
+    ];
+
+    assert_eq_and_type!(m, res);
+}
+
+#[test]
+fn stack_dynamic() {
+    let m = stack![
+        matrix![ 1, 2; 3, 4; ], 0;
+        0, dmatrix![7, 8, 9; 10, 11, 12; ];
+    ];
+
+    let res = dmatrix![
+        1, 2, 0, 0, 0;
+        3, 4, 0, 0, 0;
+        0, 0, 7, 8, 9;
+        0, 0, 10, 11, 12;
+    ];
+
+    assert_eq_and_type!(m, res);
+}
+
+#[test]
+fn stack_nested() {
+    let m = stack![
+        stack![ matrix![1, 2; 3, 4;]; matrix![5, 6;]],
+        stack![ matrix![7;9;10;], matrix![11; 12; 13;] ];
+    ];
+
+    let res = matrix![
+        1, 2, 7, 11;
+        3, 4, 9, 12;
+        5, 6, 10, 13;
+    ];
+
+    assert_eq_and_type!(m, res);
+}
+
+#[test]
+fn stack_single() {
+    let a = matrix![1, 2; 3, 4];
+    let b = stack![a];
+
+    assert_eq_and_type!(a, b);
+}
+
+#[test]
+fn stack_single_row() {
+    let a = matrix![1, 2; 3, 4];
+    let m = stack![a, a];
+
+    let res = matrix![
+        1, 2, 1, 2;
+        3, 4, 3, 4;
+    ];
+
+    assert_eq_and_type!(m, res);
+}
+
+#[test]
+fn stack_single_col() {
+    let a = matrix![1, 2; 3, 4];
+    let m = stack![a; a];
+
+    let res = matrix![
+        1, 2;
+        3, 4;
+        1, 2;
+        3, 4;
+    ];
+
+    assert_eq_and_type!(m, res);
+}
+
+#[test]
+fn stack_expr() {
+    let a = matrix![1, 2; 3, 4];
+    let b = matrix![5, 6; 7, 8];
+    let m = stack![a + b; b - a];
+
+    let res = matrix![
+        6, 8;
+        10, 12;
+        4, 4;
+        4, 4;
+    ];
+
+    assert_eq_and_type!(m, res);
+}
+
+#[test]
+fn stack_trybuild_tests() {
+    let t = trybuild::TestCases::new();
+
+    // Verify error message when try to conactenate no matrices
+    t.compile_fail("tests/trybuild/stack_empty.rs");
+    t.compile_fail("tests/trybuild/stack_empty_row.rs");
+    t.compile_fail("tests/trybuild/stack_empty_col.rs");
+}
```

Comment:
Would it be a good idea to add to this file tests to ensure that the macro doesn't explode if it's passed a matrix with zero rows or zero columns, e.g.
```rust
stack![
  a, 0;
  0, b;
];
```
Should yield a result equal to `a` if `b` is a dynamic matrix with 0 rows and 0 columns.

Actually, that's something that might be worth considering -- what should the behaviour of this macro be if passed a `DMatrix` with zero in one dimension and nonzero in another dimension, and then asked to use `0`? E.g., the above, but the matrix `b` has 2 rows and 0 columns. If `a` were `[1, 2; 3, 4]` then I think the most reasonable result would be
```rust
[
  1, 2;
  3, 4;
  0, 0;
  0, 0;
]
```
as that's the most predictable result of having that row and column count specified.

This also raises a question of whether a dynamic matrix with 3 rows and 0 columns should be considered "compatible" with a stack row that contains matrices with 5 rows. In theory it could be, although again raising an error might be the most predictable behaviour, as it lets us "take the dynamic matrix at its word" that it contains precisely 3 rows, even if it's got no entries.